bionemo-recipes/models/esm2/tests/test_thd.py

```diff
@@ -90,19 +96,101 @@ def test_thd_values_match(te_model_checkpoint, tokenizer):
-        # but it's missing this THD implementation.
-        monkeypatch.setenv("NVTE_FUSED_ATTN", "0")
+def test_thd_backwards_works(te_model_checkpoint, input_data_thd, attn_impl):
+    if attn_impl == "fused_attn" and torch.cuda.get_device_capability() == (12, 0):
+        pytest.xfail("BIONEMO-2840: On sm120 the THD backwards implementation is not available for fused attn.")
 
     model_thd = NVEsmForMaskedLM.from_pretrained(te_model_checkpoint, attn_input_format="thd", dtype=torch.bfloat16)
     model_thd.to("cuda")
     input_data = {k: v.to("cuda") if isinstance(v, torch.Tensor) else v for k, v in input_data_thd.items()}
     outputs = model_thd(**input_data)
     outputs.loss.backward()
+
+
+def test_thd_backwards_passes_match(te_model_checkpoint, tokenizer, test_proteins, attn_impl):
+    if attn_impl == "fused_attn" and torch.cuda.get_device_capability() == (12, 0):
+        pytest.xfail("BIONEMO-2840: On sm120 the THD backwards implementation is not available for fused attn.")
+
+    sequences = [tokenizer(protein) for protein in test_proteins]
+    sequences = [
+        {
+            "input_ids": seq["input_ids"],
+            "attention_mask": seq["attention_mask"],
+            "labels": seq["input_ids"],
+        }
+        for seq in sequences
+    ]
+
+    bshd_collator = DataCollatorForTokenClassification(tokenizer=tokenizer, padding=True)
+    thd_collator = DataCollatorWithFlattening()
+
+    input_data_bshd = bshd_collator(sequences)
+    input_data_thd = thd_collator(sequences)
+
+    model_bshd = NVEsmForMaskedLM.from_pretrained(te_model_checkpoint, dtype=torch.bfloat16)
+    model_thd = NVEsmForMaskedLM.from_pretrained(te_model_checkpoint, attn_input_format="thd", dtype=torch.bfloat16)
+    model_bshd.to("cuda")
+    model_thd.to("cuda")
+
+    input_data_bshd = {k: v.to("cuda") for k, v in input_data_bshd.items()}
+    input_data_thd = {k: v.to("cuda") if isinstance(v, torch.Tensor) else v for k, v in input_data_thd.items()}
+
+    bshd_outputs = model_bshd(**input_data_bshd)
+    thd_outputs = model_thd(**input_data_thd)
+
+    bshd_outputs = model_bshd(**input_data_bshd)
+    thd_outputs = model_thd(**input_data_thd)
+
+    thd_outputs.loss.backward()
+    bshd_outputs.loss.backward()
+
+    thd_grads = {name: p.grad for name, p in model_thd.named_parameters() if p.grad is not None}
+    bshd_grads = {name: p.grad for name, p in model_bshd.named_parameters() if p.grad is not None}
+
+    # For some reason, the word embeddings grads have a slightly higher numerical error.
+    thd_word_embeddings_grad = thd_grads.pop("esm.embeddings.word_embeddings.weight")
+    bshd_word_embeddings_grad = bshd_grads.pop("esm.embeddings.word_embeddings.weight")
+
+    torch.testing.assert_close(thd_grads, bshd_grads)
+
+    # sus
+    torch.testing.assert_close(thd_word_embeddings_grad, bshd_word_embeddings_grad, atol=1e-3, rtol=1e-5)
```

Comment:
well, these tests all pass on H100s. Probably need to dig into this line a bit more

Reply:
yeah we could try something like that; run it for a few iterations